hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/QuotaRepairTask.java

```diff
@@ -64,117 +72,188 @@ public class QuotaRepairTask {
+      } catch (Exception exp) {
+        LOG.error("failed to cleanup", exp);
+      }
+      IN_PROGRESS.set(false);
     }
-    updateOldVolumeQuotaSupport();
-
-    // cleanup epoch added to avoid extra epoch id in cache
-    ArrayList<Long> epochs = new ArrayList<>();
-    epochs.add(EPOCH_DEFAULT);
-    metadataManager.getBucketTable().cleanupCache(epochs);
-    metadataManager.getVolumeTable().cleanupCache(epochs);
+    return true;
   }
-  
-  private void prepareAllVolumeBucketInfo() throws IOException {
-    try (TableIterator<String, ? extends Table.KeyValue<String, OmVolumeArgs>>
-        iterator = metadataManager.getVolumeTable().iterator()) {
 
-      OmVolumeArgs omVolumeArgs;
-      while (iterator.hasNext()) {
-        Table.KeyValue<String, OmVolumeArgs> entry =
-            iterator.next();
-        omVolumeArgs = entry.getValue();
-        getAllBuckets(omVolumeArgs.getVolume(), omVolumeArgs.getObjectID());
-        if (omVolumeArgs.getQuotaInBytes() == OLD_QUOTA_DEFAULT
-            || omVolumeArgs.getQuotaInNamespace() == OLD_QUOTA_DEFAULT) {
-          oldVolumeKeyNameMap.put(entry.getKey(), entry.getValue().getVolume());
-        }
+  private void repairActiveDb(
+      OMMetadataManager metadataManager,
+      OzoneManagerProtocolProtos.QuotaRepairRequest.Builder builder) throws Exception {
+    Map<String, OmBucketInfo> nameBucketInfoMap = new HashMap<>();
+    Map<String, OmBucketInfo> idBucketInfoMap = new HashMap<>();
+    Map<String, OmBucketInfo> oriBucketInfoMap = new HashMap<>();
+    prepareAllBucketInfo(nameBucketInfoMap, idBucketInfoMap, oriBucketInfoMap, metadataManager);
+
+    repairCount(nameBucketInfoMap, idBucketInfoMap, metadataManager);
+
+    // prepare and submit request to ratis
+    for (Map.Entry<String, OmBucketInfo> entry : nameBucketInfoMap.entrySet()) {
+      OmBucketInfo oriBucketInfo = oriBucketInfoMap.get(entry.getKey());
+      OmBucketInfo updatedBuckedInfo = entry.getValue();
+      boolean oldQuota = oriBucketInfo.getQuotaInBytes() == OLD_QUOTA_DEFAULT
+          || oriBucketInfo.getQuotaInNamespace() == OLD_QUOTA_DEFAULT;
+      if (!(oldQuota || isChange(oriBucketInfo, updatedBuckedInfo))) {
+        continue;
       }
+      OzoneManagerProtocolProtos.BucketQuotaCount.Builder bucketCountBuilder
+          = OzoneManagerProtocolProtos.BucketQuotaCount.newBuilder();
+      bucketCountBuilder.setVolName(updatedBuckedInfo.getVolumeName());
+      bucketCountBuilder.setBucketName(updatedBuckedInfo.getBucketName());
+      bucketCountBuilder.setDiffUsedBytes(updatedBuckedInfo.getUsedBytes() - oriBucketInfo.getUsedBytes());
+      bucketCountBuilder.setDiffUsedNamespace(
+          updatedBuckedInfo.getUsedNamespace() - oriBucketInfo.getUsedNamespace());
+      bucketCountBuilder.setSupportOldQuota(oldQuota);
+      builder.addBucketCount(bucketCountBuilder.build());
     }
+
+    // update volume to support quota
+    builder.setSupportVolumeOldQuota(true);
```

Comment:
If it's always true, then suggest to remove this flag in the proto definition.

Reply:
We have a plan to trigger at bucket level, in that case, this is required. In this case, may not have usecase, but planning to have control with CLI option.